Contracts: use compiled rust tests

pgherveou opened the pull request:
see #2189

This PR does the following:
- Bring the user api functions into a new pallet-contracts-uapi (They are currently defined in ink! [here])(https://github.com/paritytech/ink/blob/master/crates/env/src/engine/on_chain/ext.rs)
- Add older api versions and unstable to the user api trait.
- Remove pallet-contracts-primitives and bring the types it defined in uapi / pallet-contracts
- Add the infrastructure to build fixtures from Rust files and test it works by replacing `dummy.wat` and `call.wat`
- Move all the doc from wasm/runtime.rs to pallet-contracts-uapi.

This will be done in a follow up:
- convert the rest of the test from .wat to rust
- bring risc-v uapi up to date with wasm
- finalize the uapi host fns, making sure everything is codegen from the source host fns in pallet-contracts

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4347728

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4349362

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4374112

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4375155

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4375553

pgherveou commented on substrate/frame/contracts/fixtures/build.rs:
```diff
@@ -0,0 +1,202 @@
+};
+use twox_hash::XxHash32;
+
+/// Salt used for hashing the contract source files.
```
```suggestion
/// Salt used for hashing contract source files.
```

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4376168

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4376313

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4376300

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4376314

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4376312

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4378265

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4379723

athei commented on substrate/frame/contracts/uapi/src/lib.rs:
```diff
@@ -0,0 +1,337 @@
+#[repr(transparent)]
+pub struct Ptr32Mut<'a, T>
+where
+	T: ?Sized,
+{
+	/// The internal Wasm32 raw pointer value.
+	///
+	/// Must not be readable or directly usable by any safe Rust code.
+	_value: u32,
+	/// We handle types like these as if the associated lifetime was exclusive.
+	marker: PhantomData<fn() -> &'a mut T>,
+}
+
+impl<'a, T> Ptr32Mut<'a, T>
+where
+	T: ?Sized,
+{
+	/// Creates a new Wasm32 pointer for the given raw pointer value.
+	fn new(value: u32) -> Self {
+		Self { _value: value, marker: Default::default() }
+	}
+}
+
+impl<'a, T> Ptr32Mut<'a, [T]> {
+	/// Creates a new Wasm32 pointer from the given exclusive slice.
+	pub fn from_slice(slice: &'a mut [T]) -> Self {
+		Self::new(slice.as_ptr() as u32)
+	}
+}
+
+impl<'a, T> Ptr32Mut<'a, T>
+where
+	T: Sized,
+{
+	/// Creates a new Wasm32 pointer from the given exclusive reference.
+	pub fn from_ref(a_ref: &'a mut T) -> Self {
+		let a_ptr: *mut T = a_ref;
+		Self::new(a_ptr as u32)
+	}
+}
```
Let's get rid of those silly wrappers. They are just noise to me. Let's just express our intention properly: With pointers. AFAIK they are only used in the wasm32 module anyways.

athei commented on substrate/frame/contracts/uapi/src/lib.rs:
```diff
@@ -0,0 +1,337 @@
+	*output = &mut tmp[..new_len];
+}
+
+// is not recognizing its allocator and panic handler definitions.
+#[cfg(not(any(feature = "std", feature = "no-allocator")))]
+mod allocator;
```
We neither need an allocator here nor should it be provided by this library. ink! brings its own. Our test fixtures will not need allocation.

athei commented on substrate/frame/contracts/uapi/src/lib.rs:
```diff
@@ -0,0 +1,337 @@
+		///    node console's `stdout` when the log level is set to `-lruntime::contracts=debug`.
+		///
+		/// # Note
+		///
+		/// This depends on the `debug_message` interface which requires the
+		/// `"pallet-contracts/unstable-interface"` feature to be enabled in the target runtime.
+		#[macro_export]
+		macro_rules! debug_print {
+			($($arg:tt)*) => ($crate::debug_message(&$crate::format!($($arg)*)));
+		}
+
+		/// Appends a formatted string to the `debug_message` buffer, as per [`debug_print`] but
+		/// with a newline appended.
+		///
+		/// # Note
+		///
+		/// This depends on the `debug_message` interface which requires the
+		/// `"pallet-contracts/unstable-interface"` feature to be enabled in the target runtime.
+		#[macro_export]
+		macro_rules! debug_println {
+			() => ($crate::debug_print!("\n"));
+			($($arg:tt)*) => (
+				$crate::debug_print!("{}\n", $crate::format!($($arg)*));
+			)
+		}
+	} else {
+		#[macro_export]
+		/// Debug messages disabled. Enable the `ink-debug` feature for contract debugging.
+		macro_rules! debug_print {
+			($($arg:tt)*) => ();
+		}
+
+		#[macro_export]
+		/// Debug messages disabled. Enable the `ink-debug` feature for contract debugging.
+		macro_rules! debug_println {
+			() => ();
+			($($arg:tt)*) => ();
+		}
+	}
+}
```
Doesn't belong in this crate. We don't do any printing here and should not be part of the uapi. If we need to do debug printing in our fixtures we need to move it there.

athei commented on substrate/frame/contracts/uapi/src/lib.rs:
```diff
@@ -0,0 +1,337 @@
+#[cfg(not(any(feature = "std", feature = "no-allocator")))]
+mod allocator;
+
+mod prelude;
```
We don't need to provide a prelude. It's not the business of the uapi.

athei commented on substrate/frame/contracts/uapi/src/wasm32.rs:
```diff
@@ -0,0 +1,681 @@
+#[repr(transparent)]
+pub struct Ptr32Mut<'a, T>
+where
+	T: ?Sized,
+{
+	/// The internal Wasm32 raw pointer value.
+	///
+	/// Must not be readable or directly usable by any safe Rust code.
+	_value: u32,
+	/// We handle types like these as if the associated lifetime was exclusive.
+	marker: PhantomData<fn() -> &'a mut T>,
+}
+
+impl<'a, T> Ptr32Mut<'a, T>
+where
+	T: ?Sized,
+{
+	/// Creates a new Wasm32 pointer for the given raw pointer value.
+	fn new(value: u32) -> Self {
+		Self { _value: value, marker: Default::default() }
+	}
+}
+
+impl<'a, T> Ptr32Mut<'a, [T]> {
+	/// Creates a new Wasm32 pointer from the given exclusive slice.
+	fn from_slice(slice: &'a mut [T]) -> Self {
+		Self::new(slice.as_ptr() as u32)
+	}
+}
+
+impl<'a, T> Ptr32Mut<'a, T>
+where
+	T: Sized,
+{
+	/// Creates a new Wasm32 pointer from the given exclusive reference.
+	fn from_ref(a_ref: &'a mut T) -> Self {
+		let a_ptr: *mut T = a_ref;
+		Self::new(a_ptr as u32)
+	}
+}
```
Let's get rid of this noise and use pointers where necessary as in the rv32 implementation.

athei commented on substrate/frame/contracts/uapi/src/prelude.rs:
Let's not have a prelude. This crate should stay laser focused on providing the uapi.

athei reviewed (changes requested):
Please keep the uapi crate laser focused on providing only the user space API as it will be imported by ink! later on. Meaning no allocator or panic handler. Those need to be provided by the binary. Common code of our fixtures should move to the fixtures crate. Possibly as a module included by each of the fixture rather than copy paste.

We should probably have a trait describing all the host functions and then just implement it  differently per architecture. This way we make sure that we don't have different signatures per architecture or forget to implement for one of them.

We should push your TODOs to a followup PR. Especially the RISC-V one. We need a special toolchain which we need to install in the CI image first.

Maybe just convert a single test in this PR so we can see that it actually works.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4385559

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4385557

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4385558

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4385690

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4385831

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4385830

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4385829

pgherveou commented on substrate/frame/contracts/fixtures/Cargo.toml:
```diff
@@ -11,6 +11,14 @@ description = "Fixtures for testing contracts pallet."
+twox-hash = "1.6.3"
+anyhow = "1.0.0"
 
 [features]
```
mmm the crate does not support no_std so it needs the std feature. it does not compile without it.

pgherveou commented:
> Please keep the uapi crate laser focused on providing only the user space API as it will be imported by ink! later on. Meaning no allocator or panic handler. Those need to be provided by the binary. Common code of our fixtures should move to the fixtures crate. Possibly as a module included by each of the fixture rather than copy paste.

Make sense, cleaned up the uapi crate, and added a common package with the panic handler.

> We should probably have a trait describing all the host functions and then just implement it differently per architecture. This way we make sure that we don't have different signatures per architecture or forget to implement for one of them.

yes need to add that, this will require to bring it a struct + a trait in each fixture though 

> 
> We should push your TODOs to a followup PR. Especially the RISC-V one. We need a special toolchain which we need to install in the CI image first.

Agreed let's merge this first part first.

> 
> Maybe just convert a single test in this PR so we can see that it actually works.

This is already doing this, the dummy.wat has been replaced and it's used in a few tests already.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4387640

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4389169

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4390439

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4428222

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4428188

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4432834

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4432978

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4474204

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4474221

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4474970

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4475245

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4475279

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4475261

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4475260

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4476360

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4477331

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4478036

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4478914

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4479455

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4480206

pgherveou commented on substrate/frame/contracts/uapi/src/api.rs:
maybe we should just scrape this one, since it's essentially the same as _v1

athei commented on substrate/frame/contracts/uapi/src/api.rs:
```diff
@@ -0,0 +1,759 @@
+	/// # Return
+	///
+	///  Returns the number of times specified contract exists on the call stack.
+	#[deprecated(note = "Unstable")]
```
I the note should be more extensive to explain it is not really a deperecation: "Unstable function. Behaviour can change without further notice. Use only for testing".

athei commented on substrate/frame/contracts/uapi/src/api.rs:
```diff
@@ -0,0 +1,759 @@
+	/// - [CalleeTrapped][`crate::ReturnErrorCode::CalleeTrapped]
+	/// - [TransferFailed][`crate::ReturnErrorCode::TransferFailed]
+	/// - [NotCallable][`crate::ReturnErrorCode::NotCallable]
+	#[deprecated(note = "Unstable, use `call_v1` instead")]
```
Should be replaced with the general warning above. I think it is kind of obvious that we can fall back to an older version.

athei commented on substrate/frame/contracts/uapi/src/api.rs:
Kind of obvious. We should state why it is deprecated instead: It is unsafe to not limit the PoV size. However, the only safe function is unstable :D

pgherveou replied:
yeah maybe we just scrape this one, this one became deprecated because re removed the deprecated  `code_hash_len` and `value_len` on the host side

athei replied:
Yes sure we can remove it.

athei reviewed (changes requested):
All the types that can be reasonably moved to `uapi` should be moved there. Especially the types used in the runtime API. Reason is that software like `cargo-contract` imported `primitives` in order to interact with the runtime APIs.  `pallet-contracts` is a much heavier dependency for to import. Ideally, they can just import `uapi`.

Also, let's remove all the random `#[inline(always)]` attributes. Without any evidence that this does anything it is just noise. Especially when they are used within the same crate.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4502515

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4502549

pgherveou commented on substrate/frame/contracts/fixtures/build.rs:
👍 done, common should be good though since it only has lib.rs that is part of the crate

athei replied:
But `common` should not be part of the workspace. Otherwise it is being build for the host. That is a problem because we want to put code in there that only compiles for contracts. Even right now this should not work as the panic handler only works on the target architectures.

Maybe we should not use a whole crate here. We can just have a module that is imported by all contracts.

athei replied:
Resolved: We just conditionally compile the whole crate.

pgherveou commented on substrate/frame/contracts/uapi/src/api.rs:
```diff
@@ -0,0 +1,759 @@
+	/// Equivalent to the newer [`Self::clear_storage_v1`] version with
+	/// the exception of the return type. Still a valid thing to call when not interested in the
+	/// return value.
+	fn clear_storage(key: &[u8]);
```
yes I think this is the only case with set_storage where we have a new version without deprecating the old one, at least that's what the original code comments are saying

pgherveou commented on substrate/frame/contracts/uapi/src/lib.rs:
```diff
@@ -0,0 +1,172 @@
+		Some(ref data) => data.as_ptr(),
+		None => SENTINEL as _,
+	}
+}
```
moved it with this commit https://github.com/paritytech/polkadot-sdk/pull/2347/commits/b55b3706ca4966165bd3d9ea0c0b6d4a8f09f11c
Not sure if that's what you had in mind. You need one way or another to have these functions behind the cfg flag or the build will complain they are never used

athei replied:
Just conditionally compile the whole `Api` module to avoid this problem. Then we don't need the `common` sub module and don't need to duplicate the `ApiImpl` type. Can just be defined once in the module root.

pgherveou commented on substrate/frame/contracts/fixtures/contracts/call.rs:
```diff
@@ -0,0 +1,48 @@
+#![no_main]
+
+extern crate common;
+extern crate uapi;
```
ah might be missing the edition when we compile the contracts, ... checking

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4509160

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4509143

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-tracing
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4509178

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4509159

athei commented on substrate/frame/contracts/fixtures/build.rs:
```diff
@@ -0,0 +1,277 @@
+[[bin]]
+
+[dependencies]
+uapi = {{ package = 'pallet-contracts-uapi', default-features = false,  path = {uapi_path:?}}}
```
No longer necessary.

```suggestion
uapi = {{ package = 'pallet-contracts-uapi', path = {uapi_path:?}}}
```

athei commented on substrate/frame/contracts/fixtures/contracts/call.rs:
```diff
@@ -0,0 +1,48 @@
+use uapi::{Api, ApiImpl as api, CallFlags};
+
+#[no_mangle]
+pub fn deploy() {}
```
```suggestion
pub extern "C" fn deploy() {}
```

athei commented on substrate/frame/contracts/fixtures/contracts/call.rs:
```diff
@@ -0,0 +1,48 @@
+pub fn deploy() {}
+
+#[no_mangle]
+pub fn call() {
```
```suggestion
pub extern "C" fn call() {
```

athei commented on substrate/frame/contracts/fixtures/contracts/dummy.rs:
```diff
@@ -0,0 +1,26 @@
+extern crate common;
+
+#[no_mangle]
+pub fn deploy() {}
```
```suggestion
pub extern "C" fn deploy() {}
```

athei commented on substrate/frame/contracts/fixtures/contracts/dummy.rs:
```diff
@@ -0,0 +1,26 @@
+pub fn deploy() {}
+
+#[no_mangle]
+pub fn call() {}
```
```suggestion
pub extern "C" fn call() {}
```

athei commented on substrate/frame/contracts/uapi/src/lib.rs:
```diff
@@ -0,0 +1,139 @@
+            $name:ident = $discr:literal,
+        )*
+    ) => {
+        /// Every error that can be returned to a contract when it calls any of the host functions.
+		#[derive(Debug)]
+        #[repr(u32)]
+        pub enum ReturnErrorCode {
+			/// API call successful.
+            Success = 0,
+            $(
+                $( #[$attr] )*
+                $name = $discr,
+            )*
+            /// Returns if an unknown error was received from the host module.
+            Unknown,
+        }
```
Still some space / tab stuff going on here probably.

athei commented on substrate/frame/contracts/uapi/src/host/wasm32.rs:
```diff
@@ -0,0 +1,812 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// #![allow(unused_variables)]
```
Some leftover?

athei commented on substrate/frame/contracts/uapi/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+paste = { version = "1.0", default-features = false }
+bitflags = "1.0"
+
+[features]
+default = [ "scale" ]
+scale = [ "dep:scale", "scale-info" ]
```
I don't really see where this feature is used? Scale is just used unconditionally to encode parameters when necessary. This is only the case in RISC-V which will be added in a follow up.

But even then it should be an architecture dependent dependency and not a feature.

pgherveou replied:
yep my bad this is for RISC-V and should be architecture dependent

athei commented on substrate/frame/contracts/uapi/src/lib.rs:
```diff
@@ -0,0 +1,139 @@
+mod flags;
+pub use flags::*;
+
+#[cfg(any(target_arch = "wasm32", target_arch = "riscv32"))]
+mod host;
+
+#[cfg(any(target_arch = "wasm32", target_arch = "riscv32"))]
+pub use host::*;
```
Nit: Can we de-duplicate those attributes somehow?

pgherveou replied:
if we bring back cfg_if, not sure it's worth it for just saving this one line

athei replied:
Yeah not worth it then.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4518678

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4520057

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4520091

athei commented on substrate/frame/contracts/uapi/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+description = "Exposes all the host functions that a contract can import."
+
+[dependencies]
+paste = { version = "1.0", default-features = false }
```
This is unused, right?

pgherveou replied:
it's used by the hash_fn  and impl_wrapper_for macro that both need to do some function name concatenation

xermicus commented on substrate/frame/contracts/fixtures/build.rs:
```diff
@@ -0,0 +1,277 @@
+
+/// Read the file at `path` and return its hash as a hex string.
+fn file_hash(path: &Path) -> String {
+	let data = fs::read(path).expect("file exists; qed");
```
Total nit, this isn't actually true but fine for me if it is enforced further up (also it's a build script and not a critical path in the runtime so not the end of the world if we panic here I guess)

xermicus reviewed (approved):
LGTM!

pgherveou commented on substrate/frame/contracts/fixtures/contracts/common/Cargo.toml:
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "pallet-contracts-fixtures-common"
+publish = false
+version = "1.0.0"
```
It does not really matter here, this is not going to be published. 
also the other versions in master are kind of useless as well, they get swapped by the latest published version (+ increment) when a new version is published

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v1-5-0/5358/1
